File: Data Structure/Hash/Medium/LET_no974_Subarray Sums Divisible by K.cpp
class Solution {
public:
    int subarraysDivByK(vector<int>& nums, int k) {
        vector<int> memo(k, 0);
        memo[0] = 1;
        int prefix_sum = 0, res = 0;
        
        for (int num : nums) {
            prefix_sum = ((prefix_sum + num) % k + k) % k;
            res += memo[prefix_sum];
            memo[prefix_sum]++;
        }

        return res;
    }
};

File: Data Structure/Hash/Medium/LET_no974_Subarray Sums Divisible by K.py
class Solution:
    def subarraysDivByK(self, nums: List[int], k: int) -> int:
        memo = [0 for _ in range(k)]
        memo[0] = 1
        res = prefix_sum = 0

        for num in nums:
            prefix_sum = ((prefix_sum + num) % k + k) % k
            res += memo[prefix_sum]
            memo[prefix_sum] += 1

        return res